flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -83,9 +88,25 @@
 
 	// ------------------------------------------------------------------------
 
-	public AbstractMetricGroup(MetricRegistry registry, String[] scope) {
+	public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
 		this.registry = checkNotNull(registry);
 		this.scopeComponents = checkNotNull(scope);
+		this.parent = parent;
+	}
+
+	public Map<String, String> getAllVariables() {
+		if (variables == null) { // avoid synchronization for common case
```

Comment:
well this is interesting. Making the access more expensive could pose problems for the JMX exposure I linked, as it calls `getAllVariables()` for every metric. Would the locking issue still apply if we initialize the maps eagerly, but insert the values lazily? Instead of a null-check we would use `Map#isEmpty()`.